wolfcrypt/src/aes.c

```diff
@@ -7379,6 +7707,1004 @@ int wc_AesGcmDecrypt(Aes* aes, byte* out, const byte* in, word32 sz,
+ * @return  AES_GCM_AUTH_E when authentication tag doesn't match calculated
+ *          value.
+ */
+static int AesGcmDecryptFinal_aesni(Aes* aes, const byte* authTag,
+                                    word32 authTagSz)
+{
+    int ret = 0;
+    int res;
+
+    SAVE_VECTOR_REGISTERS();
+    if (aes->cOver > 0) {
+        /* Zeroize the unused part of the block. */
+        XMEMSET(aes->lastBlock + aes->cOver, 0, AES_BLOCK_SIZE - aes->cOver);
+        /* Hash the last block of cipher text. */
+    #ifdef HAVE_INTEL_AVX2
+        if (IS_INTEL_AVX2(intel_flags)) {
+            AES_GCM_ghash_block_avx2(aes->lastBlock, aes->tag, aes->H);
+        }
+        else
+    #endif
+    #ifdef HAVE_INTEL_AVX1
+        if (IS_INTEL_AVX1(intel_flags)) {
+            AES_GCM_ghash_block_avx1(aes->lastBlock, aes->tag, aes->H);
+        }
+        else
+    #endif
+        {
+            AES_GCM_ghash_block_aesni(aes->lastBlock, aes->tag, aes->H);
+        }
+    }
+    /* Calculate and compare the authentication tag. */
+#ifdef HAVE_INTEL_AVX2
+    if (IS_INTEL_AVX2(intel_flags)) {
+        AES_GCM_decrypt_final_avx2(aes->tag, authTag, authTagSz, aes->cSz,
+            aes->aSz, aes->H, aes->initCtr, &res);
+    }
+    else
+#endif
+#ifdef HAVE_INTEL_AVX1
+    if (IS_INTEL_AVX1(intel_flags)) {
+        AES_GCM_decrypt_final_avx1(aes->tag, authTag, authTagSz, aes->cSz,
+            aes->aSz, aes->H, aes->initCtr, &res);
+    }
+    else
+#endif
+    {
+        AES_GCM_decrypt_final_aesni(aes->tag, authTag, authTagSz, aes->cSz,
+            aes->aSz, aes->H, aes->initCtr, &res);
+    }
+    RESTORE_VECTOR_REGISTERS();
+    /* Return error code when calculated doesn't match input. */
+    if (res == 0) {
+        ret = AES_GCM_AUTH_E;
+    }
+    return ret;
+}
+#endif /* HAVE_AES_DECRYPT || HAVE_AESGCM_DECRYPT */
+#endif /* WOLFSSL_AESNI */
+
+/* Initialize an AES GCM cipher for encryption or decryption.
+ *
+ * @param [in, out] aes   AES object.
+ * @param [in]      key   Buffer holding key.
+ * @param [in]      len   Length of key in bytes.
+ * @param [in]      iv    Buffer holding IV/nonce.
+ * @param [in]      ivSz  Length of IV/nonce in bytes.
+ * @return  0 on success.
+ * @return  BAD_FUNC_ARG when aes is NULL, or a length is non-zero but buffer
+ *          is NULL, or the IV is NULL and no previous IV has been set.
+ */
+int wc_AesGcmInit(Aes* aes, const byte* key, word32 len, const byte* iv,
+    word32 ivSz)
+{
+    int ret = 0;
+
+    /* Check validity of parameters. */
+    if ((aes == NULL) || ((len > 0) && (key == NULL)) || ((ivSz > 0) &&
+            (iv == NULL))) {
+        ret = BAD_FUNC_ARG;
+    }
+
+    /* Set the key if passed in. */
+    if ((ret == 0) && (key != NULL)) {
+        ret = wc_AesGcmSetKey(aes, key, len);
+    }
+
+    if (ret == 0) {
+        /* Setup with IV if needed. */
+        if (iv != NULL) {
+            /* Cache the IV in AES GCM object. */
+            XMEMCPY((byte*)aes->reg, iv, ivSz);
```

Comment:
Check if ivSz is valid (e.g. `<= sizeof(aes->reg)`) to prevent OOB write.

Reply:
Push a fix for this. Checking IV size is not greater than AES_BLOCK_SIZE.

Reply:
Tested and confirmed fixed.

Please also reject IV of size 0, as the spec prohibits this.

Reply:
Done!